### PR TITLE
Add LLVM Golden IRs

### LINKED ISSUE #614: Debug tooling support for LLVM IR
- [ ] view in VS Code (`/vscode_extension/src/debug_irs.ts`)
- [ ] view via CLI (`/compiler/cli/src/debug.rs`)
- [ ] include in golden files (`/compiler/cli/src/debug.rs`)

### LINKED ISSUE #649: Golden support for LLVM bytecode

### PR DESCRIPTION
**Pull request by ctiedt**
Closes: #649 
Closes: #614 

### Checklist

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,6 +2,7 @@ generated/
 local/
 
 *.candy.*
+**/.goldens/
```
Comment (JonasWanke) [inline .gitignore]:
```suggestion
.goldens/
```

File: compiler/backend_inkwell/src/lib.rs [hunk 2]
```diff
@@ -11,14 +16,58 @@ use inkwell::{
...
 
+#[salsa::query_group(LlvmIrStorage)]
+pub trait LlvmIrDb: OptimizeMir {
+    #[salsa::transparent]
+    fn llvm_ir(&self, module: CandyModule) -> Result<RichIr, ModuleError>;
+}
+
+#[allow(clippy::needless_pass_by_value)]
+fn llvm_ir(db: &dyn LlvmIrDb, module: CandyModule) -> Result<RichIr, ModuleError> {
+    //debug!("{module}: Compiling.");
```
Comment (JonasWanke) [inline compiler/backend_inkwell/src/lib.rs]:
```suggestion
```

File: compiler/backend_inkwell/src/lib.rs [hunk 3]
```diff
@@ -11,14 +16,58 @@ use inkwell::{
...
+            println!("{:?}", error);
+        }
+        std::process::exit(1);
+    }
+
+    let context = Context::create();
+    let mut codegen = CodeGen::new(&context, "module", mir);
+    let llvm_ir = codegen.compile("", false, true).unwrap();
+
+    //debug!("{module}: Done. Optimized from {complexity_before} to {complexity_after}");
```
Comment (JonasWanke) [inline compiler/backend_inkwell/src/lib.rs]:
```suggestion
```

File: compiler/backend_inkwell/src/lib.rs [hunk 4]
```diff
@@ -11,14 +16,58 @@ use inkwell::{
...
+                FxHashSet::from_iter([CompilerError::for_whole_module(module.clone(), payload)]);
+            (Arc::new(mir), Arc::new(errors))
+        });
+
+    if !errors.is_empty() {
+        for error in errors.iter() {
+            println!("{:?}", error);
+        }
+        std::process::exit(1);
+    }
```
Comment (JonasWanke) [inline compiler/backend_inkwell/src/lib.rs]: I think you can just do this:

```suggestion
    let (mir, errors) = db.optimized_mir(module.clone(), TracingConfig::off())?;
```

File: compiler/backend_inkwell/src/lib.rs [hunk 5]
```diff
@@ -223,9 +272,11 @@ impl<'ctx> CodeGen<'ctx> {
             self.module.print_to_stderr();
         }
         self.module.verify()?;
-        let bc_path = PathBuf::from(format!("{path}.bc"));
-        self.module.write_bitcode_to_path(&bc_path);
-        Ok(())
+        if !path.is_empty() {
```
Comment (JonasWanke) [inline compiler/backend_inkwell/src/lib.rs:250]: Maybe rather accept an `Option<&str>`?

File: compiler/cli/src/debug.rs [hunk 6]
```diff
@@ -58,6 +60,10 @@ pub(crate) enum Options {
     /// Low-Level Intermediate Representation
     Lir(PathAndTracing),
 
+    /// LLVM Intermediate Representation
+    #[cfg(feature = "inkwell")]
+    LlvmIr(PathAndTracing),
+
```
Comment (JonasWanke) [inline compiler/cli/src/debug.rs]: Please move this below the VM's byte code

File: compiler/language_server/src/features_ir.rs [hunk 7]
```diff
@@ -116,6 +118,8 @@ impl IrFeatures {
                 &db.lir(config.module.clone(), tracing_config.clone()),
                 tracing_config,
             ),
+            #[cfg(feature = "inkwell")]
+            Ir::LlvmIr => db.llvm_ir(config.module.clone()).unwrap(),
```
Comment (JonasWanke) [inline compiler/language_server/src/features_ir.rs]: Please move this below the VM's byte code

File: vscode_extension/src/debug_irs.ts [hunk 8]
```diff
@@ -10,6 +10,7 @@ type Ir =
   | { type: "mir"; tracingConfig: TracingConfig }
   | { type: "optimizedMir"; tracingConfig: TracingConfig }
   | { type: "lir"; tracingConfig: TracingConfig }
+  | { type: "llvmIr" }
```
Comment (JonasWanke) [inline vscode_extension/src/debug_irs.ts]: Please move this below the VM's byte code

File: vscode_extension/package.json [hunk 9]
```diff
@@ -141,6 +141,11 @@
         "command": "candy.debug.viewOptimizedMir",
         "title": "View Optimized MIR"
       },
+      {
+        "category": "Candy Compiler Debugging",
+        "command": "candy.debug.viewLlvmIr",
+        "title": "View LLVM IR"
+      },
```
Comment (JonasWanke) [inline vscode_extension/package.json]: Please move this below the VM's byte code

File: compiler/cli/src/debug.rs [hunk 10]
```diff
@@ -58,6 +60,10 @@ pub(crate) enum Options {
     /// Low-Level Intermediate Representation
     Lir(PathAndTracing),
 
+    /// LLVM Intermediate Representation
+    #[cfg(feature = "inkwell")]
+    LlvmIr(PathAndTracing),
```
Comment (JonasWanke) [inline compiler/cli/src/debug.rs]:
```suggestion
    LlvmIr(OnlyPath),
```

**Review by JonasWanke — CHANGES_REQUESTED** [review]

File: compiler/backend_inkwell/src/lib.rs (hunk 5 above)
Comment (ctiedt) [reply compiler/backend_inkwell/src/lib.rs:250]: This will be changed when I work on #717

**Review by JonasWanke — APPROVED** [review]

**Pull request merged**